src/dict.c

```diff
@@ -509,6 +562,8 @@ dictEntry *dictInsertAtPosition(dict *d, void *key, void *position) {
             /* Allocate an entry without value. */
             entry = createEntryNoValue(key, *bucket);
         }
+    } else if (d->type->embedded_entry) {
+        entry = createEmbeddedEntry(key, *bucket, d->type);
```

Comment:
If the dict type has no key dup callback, then `dictAddRaw()` used to take ownership of the key. This semantics is changed here if embedded key is enabled for the dict type. Then we copy the key instead. This means that the caller needs to free the key after calling `dictAddRaw()`.

That's why you needed to insert an `sdsfree()` call in rdb.c, I believe.

I suppose it affects `kvstoreDictAddRaw()` too, since it's calling dictAddRaw, and in turn, all functions calling kvstoreDictAddRaw, which are dbAddInternal, dbAddRDBLoad and setExpire.

I don't know if that's all. Maybe there are more...

pubsubSubscribeChannel calls kvstoreDictAddRaw too, but in this case, the dicts aren't using embedded entries (right?), so I suppose kvstoreDictAddRaw (and dictAddRaw) still takes ownership of the key in this case. (Maybe we should use embedded key for pubsub channels dicts too, so we can document that kvstoreDictAddRaw always copies the key?)

Short story: It becomes a mess. :)

We should at least document this behaviour at dictAdd, dictAddRaw and kvstoreDictAddRaw.

Reply:
There are multiple levels of API to add data to the dictionary and would be lot of duplication of code comment. Should we add more documentation at the dictType level itself?

Reply:
Duplication of documentation is okay, since that is what people will actually refer to. I think we should update the documentation in the three places viktor mentioned.

Reply:
Sure, we can add some docs for dictType too, but I'm not sure what we should write there.

The caller of e.g. `pubsubSubscribeChannel()` may not know (and shouldn't need to know) that it using dictAddRaw and that the dictType has no val dup function, so it's better to document each function separately.

Reply:
Let me know if the wording looks correct.

https://github.com/valkey-io/valkey/pull/541/commits/ce76a16ef40e81e7a0c08dd902f7961dbcaa3da0